Document Test-Driven Workflow in CONTRIBUTING.md

## Summary

The current `CONTRIBUTING.md` has a basic "Contributing Code" section but lacks detailed guidance on the **test-driven bug fix / feature workflow** that we expect contributors to follow. This workflow ensures PRs are convincing beyond any doubt and prevents regressions.

## Problem

Contributors (especially new ones) often submit PRs that:
- Don't include tests proving the bug exists
- Don't demonstrate the fix actually works
- Don't provide evidence from both local runs AND CI

This makes review harder and less certain.

## Proposed Documentation

Add a new section **"Bug Fix Workflow"** to `CONTRIBUTING.md` documenting the expected process:

### Bug Fix Workflow (Test-Driven)

When fixing a bug, follow this workflow to create a convincing, reviewable PR:

#### 1. File Issue & Discuss

- File a GitHub issue describing the bug
- Discuss and gather agreement: "yes, current behavior _is_ wrong"
- Reference any related issues or user reports

#### 2. Create PR with Failing Test

- Create a feature branch
- Write a test that **demonstrates the bug** (test should FAIL)
- Include any audit/fixture files needed
- Push the PR

#### 3. Prove the Bug Exists

- Run tests locally → verify they FAIL
- Let GitHub Actions run → verify CI FAILS
- **Leave a PR comment** with:
  - Your local test output showing failure
  - Link to the failing CI workflow run

#### 4. Implement the Fix

- Fix the bug
- Push the fix to the same PR

#### 5. Prove the Fix Works

- Run tests locally → verify they PASS
- Let GitHub Actions run → verify CI PASSES
- **Leave a PR comment** with:
  - Your local test output showing success
  - Link to the passing CI workflow run

#### 6. Request Review

- Comment: "Ready for review & merge!"
- Ping maintainers if needed

### Why This Workflow?

This leaves maintainers with a **completely convincing PR** because:

- There **is** a behavior considered incorrect (bug)
- The test **proves** the current implementation is behaving incorrectly
- The fix **proves** the new implementation does fix it
- The test coverage ensures we **won't regress** later

### Example PR Comments

**After step 3 (proving bug exists):**
```
## Bug Confirmed

Local test run:
​```
$ just test
...
FAILED test_foo.py::test_bar - AssertionError: expected X, got Y
​```

CI also fails: https://github.com/crossbario/crossbar/actions/runs/12345
```

**After step 5 (proving fix works):**
```
## Fix Verified

Local test run:
​```
$ just test
...
PASSED test_foo.py::test_bar
​```

CI passes: https://github.com/crossbario/crossbar/actions/runs/12346

Ready for review & merge!
```

## Scope

This documentation should be added to `CONTRIBUTING.md` in all Crossbar.io related Python repos: txaio, autobahn-python, zlmdb, cfxdb, wamp-xbr

## Related

- #2133 (discussion where this workflow was clarified)

## Checklist

- [x] I have searched existing issues to avoid duplicates
- [x] I have provided a minimal reproducible example
- [x] I have included version information
- [x] I have included error messages/logs

--- a/.audit/oberstet_fix_2147_2152.md
+++ b/.audit/oberstet_fix_2147_2152.md
@@ -1,0 +1,8 @@
+- [ ] I did **not** use any AI-assistance tools to help create this pull request.
+- [x] I **did** use AI-assistance tools to *help* create this pull request.
+- [x] I have read, understood and followed the projects' [AI Policy](https://github.com/crossbario/autobahn-python/blob/main/AI_POLICY.md) when creating code, documentation etc. for this pull request.
+
+Submitted by: @oberstet
+Date: 2026-06-21
+Related issue(s): #2147, #2152
+Branch: oberstet:fix_2147_2152

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -38,14 +38,44 @@ When proposing a new **feature**, please provide:
 
 ## Contributing Code
 
-1. **Fork the repository** on GitHub
-2. **Create a feature branch** from `master`
-3. **Make your changes** following the code style
-4. **Add tests** for new functionality
-5. **Run the test suite** to ensure nothing is broken
-6. **Submit a pull request** referencing any related issues
+1. **File or find an issue** first, and get agreement that the change is wanted
+2. **Fork the repository** on GitHub
+3. **Create a feature branch** from `master`
+4. **Make your changes** following the code style
+5. **Add tests** that prove your change (see [Bug Fix Workflow (Test-Driven)](#bug-fix-workflow-test-driven))
+6. **Run the test suite** locally to ensure nothing is broken
+7. **Add an AI-assistance disclosure** audit file (see [AI-Assistance Disclosure](#ai-assistance-disclosure))
+8. **Submit a pull request** referencing the originating issue
 
 We use the Fork & Pull Model. This means that you fork the repo, make changes to your fork, and then make a pull request here on the main repo.
+
+### AI-Assistance Disclosure
+
+Every pull request must include an **audit file** declaring whether AI-assistance
+tools (e.g. GitHub Copilot, Claude, ChatGPT, Cursor) were used to help create it.
+
+Add a file at `.audit/<github-username>_<branch>.md` (for example
+`.audit/jane_fix-1234.md`) containing:
+
+```markdown
+- [ ] I did **not** use any AI-assistance tools to help create this pull request.
+- [x] I **did** use AI-assistance tools to *help* create this pull request.
+- [x] I have read, understood and followed the project's [AI Policy](https://github.com/crossbario/autobahn-python/blob/main/AI_POLICY.md) when creating code, documentation etc. for this pull request.
+
+Submitted by: @<github-username>
+Date: <YYYY-MM-DD>
+Related issue(s): #<issue-number>
+Branch: <github-username>:<branch>
+```
+
+Tick the box that applies. The `Related issue(s):` line must reference the
+issue the PR addresses.
+
+> **Filename:** use `<github-username>_<branch>.md` with an **underscore**, and
+> only cross-platform-safe characters (`A-Z a-z 0-9 . _ -`). Do **not** copy the
+> GitHub `owner:branch` label literally — a `:` (or `/`) in the filename breaks
+> `git checkout` on Windows. The `:` belongs only *inside* the file, on the
+> `Branch:` line.
 
 ### Contributor Assignment Agreement
 
@@ -64,23 +94,121 @@ The CAA gives us the rights to your code, which we need e.g. to react to license
 
 **You only need to do this once - all future contributions are covered!**
 
+## Bug Fix Workflow (Test-Driven)
+
+When fixing a bug, follow this test-driven workflow to create a convincing,
+reviewable PR that proves both the bug *and* the fix, and guards against
+regressions:
+
+### 1. File the issue & agree
+
+- File a GitHub issue describing the bug.
+- Discuss and reach agreement: "yes, the current behavior *is* wrong".
+- Reference any related issues or user reports.
+
+### 2. Open a PR with a failing test
+
+- Create a feature branch.
+- Write a test that **demonstrates the bug** (the test must FAIL).
+- Include any audit/fixture files needed.
+- Push the PR.
+
+### 3. Prove the bug exists (red)
+
+- Run the tests locally → verify they FAIL.
+- Let GitHub Actions run → verify CI FAILS.
+- **Leave a PR comment** with your local failure output **and** a link to the
+  failing CI run.
+
+### 4. Implement the fix
+
+- Fix the bug.
+- Push the fix to the same PR.
+
+### 5. Prove the fix works (green)
+
+- Run the tests locally → verify they PASS.
+- Let GitHub Actions run → verify CI PASSES.
+- **Leave a PR comment** with your local success output **and** a link to the
+  passing CI run.
+
+### 6. Request review
+
+- Comment: "Ready for review & merge!" and ping a maintainer if needed.
+
+### Why this workflow?
+
+It leaves maintainers with a **completely convincing PR**, because:
+
+- there *is* a behavior considered incorrect (the bug);
+- the test *proves* the current implementation behaves incorrectly;
+- the fix *proves* the new implementation resolves it;
+- the test stays in the suite, so we **won't regress** later.
+
+The same red→green discipline applies to features: write the test for the new
+behavior first (red), then implement until it passes (green).
+
+### Example PR comments
+
+After step 3 (bug confirmed):
+
+````markdown
+## Bug confirmed
+
+Local test run:
+
+```
+$ just test
+...
+FAILED test_foo.py::test_bar - AssertionError: expected X, got Y
+```
+
+CI also fails: https://github.com/crossbario/crossbar/actions/runs/<id>
+````
+
+After step 5 (fix verified):
+
+````markdown
+## Fix verified
+
+Local test run:
+
+```
+$ just test
+...
+PASSED test_foo.py::test_bar
+```
+
+CI passes: https://github.com/crossbario/crossbar/actions/runs/<id>
+
+Ready for review & merge!
+````
+
 ## Development Setup
+
+Crossbar.io uses [`just`](https://github.com/casey/just) to drive development
+tasks (run `just` to list all recipes):
 
 ```bash
 git clone https://github.com/crossbario/crossbar.git
 cd crossbar
-pip install -e .[dev]
+# create a managed virtualenv and install crossbar + dev/test tooling
+just install-dev
 ```
 
 ## Running Tests
 
 ```bash
-# Run all tests
-tox
+# Run the full test suite
+just test
 
-# Run tests for specific Python version
-tox -e py312
+# Run tests for a specific Python (e.g. CPython 3.12, PyPy 3.11)
+just test cpy312
+just test pypy311
 ```
+
+Crossbar.io must run on both **CPython** and **PyPy** - run the relevant tests on
+both before submitting.
 
 ## Code Style
 

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -8,6 +8,7 @@ Changelog
 26.6.1
 ------
 
+* docs: document the test-driven Bug Fix Workflow and the PR AI-assistance disclosure (audit file) requirement in ``CONTRIBUTING.md``; refresh stale ``tox`` references to ``just`` (`#2152 <https://github.com/crossbario/crossbar/issues/2152>`_)
 * fix: ``crossbar version`` now reports the LMDB version from the binding vendored in zlmdb (the line was blank on clean installs, which have no top-level ``lmdb``); LMDB is now shown nested under zLMDB (`#2156 <https://github.com/crossbario/crossbar/issues/2156>`_)
 * new: ``crossbar version`` reports the vendored FlatBuffers version for both data-in-transit (autobahn) and data-at-rest (zlmdb), nested under their respective packages, so an admin can verify which version is active and that both agree (`#2156 <https://github.com/crossbario/crossbar/issues/2156>`_)
 * new: per-release codenames (e.g. Crossbar.io 26.6.1 — "Resistance Is Futile"), shown in the ``crossbar version`` banner (`#2227 <https://github.com/crossbario/crossbar/issues/2227>`_)

